Issue 3529

schlenkibus opened the pull request:
closes #3529

MatthiasSeeber reviewed (commented):
quick ui review:
- [x] the logic seems okay and corresponds with the hwui arrow logic
- [ ] when changing the VoiceGroup in the WebUI (top-center selector or dual-sound part-selector in sound-tab), often the Parameter Screen on the HWUI is invoked (observed with Split/Layer Sounds)
- [ ] (enhancement) the FX box shapes give me the impression they want to be clicked on (which can be useful in Single Sounds, not having to move to the top-center switch)

schlenkibus commented:
@hhoegelo thanks again for the feedback, the 4 image approach seems to work like intended

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/presenters/EditBufferPresenterProvider.java:
```diff
@@ -224,6 +228,361 @@ private void bruteForce() {
+        return fb_mix_fx != 0 && fb_mix_lvl > 0 && isFeedbackUsed(VoiceGroup.I);
+    }
+
+    public boolean calculateLayerIToFXI() {
```
Just an idea, don't know if this destroys the mental model you created for the problem: Could we reduce complexity and code size by only keep the branches that return true and return false as default? This being asked for all the boolean functions here?

hhoegelo replied:
Question still valid

schlenkibus replied:
i did it

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/world/overlay/belt/sound/SerialArrow.java:
```diff
@@ -0,0 +1,38 @@
+		setVisible(false);
+
+		EditBufferPresenterProvider.get().onChange(ebp -> {
+			switch(ebp.serialFX) {
```
Nitpick: You actually don't need to have member state, you could just invalidate in onChange(), and query the presenter (and apply the switch logic) in getSelectedPhase().

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/world/overlay/belt/sound/SingleSoundLayout.java:
```diff
@@ -1,11 +1,246 @@
+
+
+		@Override
+		public void draw(Context2d ctx, int i) {
```
unnecessary override

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/world/overlay/belt/sound/SingleSoundLayout.java:
```diff
@@ -1,11 +1,246 @@
+	}
+
+	@Override
+	public void draw(Context2d ctx, int invalidationFlags)
```
unnecessary override

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/world/overlay/belt/sound/SplitSoundLayout.java:
```diff
@@ -33,14 +34,268 @@ protected SplitSoundLayout(OverlayLayout parent) {
+		}
+	}
+
+	private class Split_I_To_Out extends SVGImage {
```
Here are some untested drafts of little helper classes to illustrate the idea:

```
class EditBufferBooleanImage extends SVGImage {
		public EditBufferBooleanImage(Control parent, Function<EditBufferPresenter, Boolean> test, String imageName) {
		super(parent, imageName);

		EditBufferPresenterProvider.get().onChange(ebp -> {
			setVisible(test.apply(ebp));
			return true;
		});
	}
}

class EditBufferIntMappedImage extends SVGImage {
	int selectedPhase = -1;

	public EditBufferIntMappedImage(Control parent, Function<EditBufferPresenter, Integer> test,
				String... imageNames) {
		super(parent, imageNames);

		EditBufferPresenterProvider.get().onChange(ebp -> {
			int newPhase = test.apply(ebp);
			if (newPhase != selectedPhase) {
				selectedPhase = newPhase;
				setVisible(selectedPhase >= 0);
				invalidate(INVALIDATION_FLAG_UI_CHANGED);
			}
			return true;
		});
	}

	@Override
	public int getSelectedPhase() {
		return selectedPhase;
	}
}

class EditBufferEnumMappedImage<E> extends SVGImage {
	int selectedPhase = -1;

	public EditBufferEnumMappedImage(Control parent, Function<EditBufferPresenter, E> test,
				Pair<E, String>... images) {
		super(parent, (String[]) Arrays.stream(images).map(x -> x.right).toArray());

		EditBufferPresenterProvider.get().onChange(ebp -> {
			E e = test.apply(ebp);
				
			int phase = 0;
				
			for (var k : images) {
				if (k.left == e) {
					setVisible(k.right != null);
					break;
				}
				phase++;
			}
			
			if (phase != selectedPhase) {
				selectedPhase = phase;
				invalidate(INVALIDATION_FLAG_UI_CHANGED);
			}
			return true;
		});
	}

	@Override
	public int getSelectedPhase() {
		return selectedPhase;
	}
}
```

hhoegelo replied:
Sorry for spamming, but I just had an idea how to generalize even more:

```
class MappedSvgImage<E> extends SVGImage {
	int selectedPhase = 0;
	Pair<E, String> imagesMap[];

	public MappedSvgImage(Control parent, Pair<E, String>... images) {
		super(parent, (String[]) Arrays.stream(images).map(x -> x.right).filter(x -> x != null).toArray());
		imagesMap = images;
	}

	public void update(E e) {
		int phase = 0;

		for (var k : imagesMap) {
			if (k.left == e) {
				setVisible(k.right != null);
				break;
			}

			if (k.right != null)
				phase++;
		}

		if (phase != selectedPhase) {
			selectedPhase = phase;
			invalidate(INVALIDATION_FLAG_UI_CHANGED);
		}
	}

	@Override
	public int getSelectedPhase() {
		return selectedPhase;
	}
}
```

Instances of this class can then be used from one single handler in the owning class like this:

```
class SomeLayout {
	SomeLayout(OverlayControl p)
	{
		super(p);

		addChild(foobar = new MappedSvgImage<Boolean>(this, new Pair(true, "ImageName.svg"), new Pair(false, null));
		addChild(foobaz = new MappedSvgImage<AnEnum>(this, new Pair(AnEnum.One, "ImageName.svg"), new Pair(AnEnum.Two, "OtherImage.svg));
		addChild(fooban = new MappedSvgImage<Integer>(this, new Pair(0, "ImageName.svg"), new Pair(6, "OtherImage.svg));

		EditBufferPresenterProvider.get().onChange(ebp -> { 
			foobar.update(ebp.memberOne);
			foobaz.update(ebp.memberTwo);
			fooban.update(ebp.memberThree);
				return true; 
		});
}
```

schlenkibus commented:
> * [ ]  when changing the VoiceGroup in the WebUI (top-center selector or dual-sound part-selector in sound-tab), often the Parameter Screen on the HWUI is invoked (observed with Split/Layer Sounds) [1](#user-content-fn-1-190fff4ba7f2cdbc877bf0fdd58c8176)
> ## Footnotes
> 1. when in the Sound Screen, changing the VoiceGroup via the WebUI may lead to a Parameter Screen (Split/Layer Sounds)

This seems to not be introduced by this PR, I checked a prior version (a derivative of main) and the behavior was present there as well. I would suggest creating a new Issue if we deem this change worthy.

schlenkibus commented:
- [ ] split fx to out pfeile klappen noch nicht
- [ ] layer fx to out pfeile klappen noch nicht

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/presenters/EditBufferPresenterProvider.java:
```diff
@@ -225,7 +230,377 @@ private void bruteForce() {
+        calculateSerialFx();
         calculateFXUnused();
+
+        notifyChanges();
```
If we notifyChanges unconditionally, we can remove a lot of logic trying to avoid unnecessary invalidations above. 
I think, it would be best to follow the pattern used above: only notifyChanges if the members actually changed their value. 
If you decide to not do it the fine grained way, then please also simplify the rest of the method so it does not do all the complicated stuff to just notifyChanges anyway in the end.

schlenkibus replied:
i decided to calculate the oldHash and a new hash after all presenter-fields are set, if the hashes differ I call notifyChanges now

hhoegelo commented on projects/web/static/nonmaps/src/main/java/com/nonlinearlabs/client/world/overlay/belt/sound/EditBufferBooleanImage.java:
```diff
@@ -0,0 +1,17 @@
+import com.nonlinearlabs.client.world.overlay.OverlayLayout;
+import com.nonlinearlabs.client.world.overlay.SVGImage;
+
+public class EditBufferBooleanImage extends SVGImage {
```
this class is actually not bound to the EditBuffer in any way, is it?

hhoegelo replied:
Is there a reason not to use MappedSvgImage<Boolean>(this, new Pair<Boolean, String>(false, null), new Pair<Boolean, String>(true, "foobar.svg")) ?

hhoegelo reviewed (commented):
As far as I can see, the X-to-T+*.svg images are not used anymore. If so, I would prefer to not add them to the repo and tell Tino that he can remove the files from his project, too.

schlenkibus commented:
the list of unused arrows i could identify:
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LB-to-RB+LB-to-RT.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-LB+LB-to-RB.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RB+LB-to-RB+LB-to-RT.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RB+LB-to-RT.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LB-to-RB+LB-to-RT.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LB-to-RB.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LB-to-RT.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LT-to-RB+LB-to-RB.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LT-to-RB+LB-to-RT+LB-to-RB.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LT-to-RB+LB-to-RT.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/LT-to-RT+LT-to-RB.svg
deleted:    projects/web/static/nonmaps/src/main/webapp/images/RT-to-LT+RB-to-LB.svg
